src/Corax/IndexWriter.cs

```diff
@@ -204,12 +225,35 @@ private unsafe void InsertAnalyzedToken(ByteStringContext context, ref IndexEntr
                         var fieldName = slice.Clone(context);
                         field[fieldName] = term = new List<long>();
                     }
-
                     AddMaybeAvoidDuplicate(term, entryId);
+
+                    if (_suggestionsTrees != null && _suggestionsTrees.TryGetValue(tokenField, out var tree))
+                    {
+                        AddSuggestions(slice, tree);
+                    }
                 }
             }
-        }        
-        
+        }
+
+        private unsafe void AddSuggestions(Slice slice, CompactTree tree)
+        {
+            var keys = NGrams.Generate(Transaction.Allocator, SuggestionTermProvider.DefaultNGramSize, slice.AsSpan(), out int keysCount).ToReadOnlySpan();
+            int keySizes = keys.Length / keysCount;
+            while (keys.Length > 0)
+            {
+                // PERF: If we keep going with this approch, the obvious optimization is to implement AddOrUpdate method to avoid
+                //       having to get the value and then re-add it. 
```

Comment:
A better approach would be to defer this operation to the `Commit()` phase.
You are very likely to get very similar ngrams, after all. So we can do all the usual processing in memory, and then handle  actually write the unique values back. This is roughly how Lucene does it, too

Reply:
Probably. But wont work for tracking removes as we lose how many there are. As insert only it would work though

Reply:
Where are you actually handling removal, however? I might have missed that?
But why would that matter? You are going to keep the data in memory with the total number you have (additions & removals).
Then add that to the stored value and remove if `0`